### PR TITLE
[TAPE] Add support for tape execution

### PR DESCRIPTION
**Pull request by josh146**
**Context:** Second PR in the tape series, following on from #792 

**Description of the Change:**

Adds support for tape execution on devices. In particular, it adds the following methods:

* `tape.execute(device, params=None)`: the user-facing call for tape execution on a given device. This is always the method users should use; if an interface is being used, this method will be wrapped to support autodifferentiation.

  Note that parameters need not be passed; if not passed, the current tape parameters will be used for execution.

* `tape.execute_device(device, params)`: a low level method for executing the tape on a device. This is where the main logic takes place. When using an interface, this will be _behind_ the interface; calling this method will not support autodifferentiation.

* `tape._execute`: an optional method (simply an alias to `tape.execute_device`). The interface may override this method if it needs to insert logic inbetween the two execute calls. It has the same signature as `execute_device`.

**Note:** the three methods, as well as their signatures, and constant getting and setting of tape parameters might seem overly complex. I arrived at this working solution after getting all three interfaces to work. In particular:

- Autograd requires that the method it wraps accepts differentiable numeric parameters as a positional argument. So it _must_ have `params` in the signature, and it _cannot_ have `device` be a positional argument!

- In TensorFlow and PyTorch, however, you want the opposite; the tape has already been constructed with tensors; you simply want to execute it as-is, and avoid needing to 're-submit' the trainable tensors to the `execute` method.

**Benefits:** Tape is now executable using all devices

**Possible Drawbacks:**

@mariaschuld pointed out that there are some potential advantages to instead having

```python
dev.execute(tape)
```

rather than

```python
tape.execute(dev)
```

be the user-facing call. However, this will require significant changes:

* The interface will need to be applied to the _device_, not the tape

* Similar interface requirements as to the `device.execute` method signature would need to be considered

* We are in a position where the interface jacobian call is in a different location in the code base to the wrapped execute method!

I think this will still be doable, but I propose we explore this _after_ the tape refactor, as it will definitely require interface + device API modifications. We may even be able to support both approaches.

**Related GitHub Issues:** n/a

### CODE REVIEW

File: pennylane/beta/tapes/tape.py [hunk 1]
```diff
@@ -532,7 +534,7 @@ def trainable_params(self):
         automatically excluded from the Jacobian computation.
 
         The number of trainable parameters determines the number of parameters passed to
-        :meth:`~.set_parameters`, :meth:`~.QuantumTape.execute`, and :meth:`~.QuantumTape.jacobian`,
+        :meth:`~.set_parameters`, :meth:`~.execute`, and :meth:`~.QuantumTape.jacobian`,
```
Comment (trbromley) [inline pennylane/beta/tapes/tape.py:537]: The link to `jacobian` will work later on right?

File: pennylane/beta/tapes/tape.py [hunk 2]
```diff
@@ -833,3 +835,105 @@ def data(self):
...
+
+        If parameters are not provided, the existing tape parameters are used:
+
+        >>> dev = qml.device("default.qubit", wires=[0, 'a'])
+        >>> tape.execute(dev)
+        array([[8.84828969e-01, 3.92449987e-03, 4.91235209e-04, 1.10755296e-01]])
+
+        Parameters can be optionally passed during execution:
+
+        >>> tape.execute(dev, params=[1.0, 0.0, 1.0])
```
Comment (trbromley) [inline pennylane/beta/tapes/tape.py:874]: I think it's obvious, but just wanted to ask the question: is it clear that the parameters of a tape don't include the choice of wires?

**Review by trbromley — APPROVED** [review]
Great!

**Comment by trbromley** [comment]
> @mariaschuld pointed out that there are some potential advantages to instead having

Is there a link to the conversation? I'm curious what the advantages are.

File: pennylane/beta/tapes/tape.py (hunk 1 above)
Comment (josh146) [reply pennylane/beta/tapes/tape.py:537]: yep! By the next PR :)

File: pennylane/beta/tapes/tape.py (hunk 2 above)
Comment (josh146) [reply pennylane/beta/tapes/tape.py:874]: I thought it was, but maybe it's not? tape parameters == tape gate parameters. They can be marked trainable or non-trainable.

Comment (trbromley) [reply pennylane/beta/tapes/tape.py:874]: Yeah I think it's ok, I just noticed it's different from writing a quantum function, whose signature can include the choice of wires as "parameters" :thinking:

**Comment by josh146** [comment]
> Is there a link to the conversation? I'm curious what the advantages are.

It was on Slack unfortunately! We didn't explore in too much detail. More so that `dev.execute(tape)` expands easily to multiple tapes, whereas `tape.execute(dev)` doesn't.

File: pennylane/beta/tapes/tape.py [hunk 3]
```diff
@@ -833,3 +835,105 @@ def data(self):
...
+            else:
+                output_dim = np.prod(res.shape)
+
+            if self.output_dim != output_dim:
+                # update the output dimension estimate with the correct value
+                self._output_dim = output_dim
+
+        except (AttributeError, TypeError):
+            # unable to determine the output dimension
+            pass
```
Comment (co9olguy) [inline pennylane/beta/tapes/tape.py:930]: codecov seems to have a legitimate complaint here

File: pennylane/beta/tapes/tape.py [hunk 4]
```diff
@@ -833,3 +835,105 @@ def data(self):
...
+
+        # restore original parameters
+        self.set_parameters(current_parameters)
+        return res
+
+    # interfaces can optionally override the _execute method
+    # if they need to perform any logic inbetween the users
```
Comment (co9olguy) [inline pennylane/beta/tapes/tape.py:937]:
```suggestion
    # if they need to perform any logic in between the user's
```

File: pennylane/beta/tapes/tape.py [hunk 5]
```diff
@@ -833,3 +835,105 @@ def data(self):
...
+
+        For more details on differentiable tape execution, see :meth:`~.execute`.
+
+        Args:
+            device (~.Device): a PennyLane device
+                that can execute quantum operations and return measurement statistics
+            params (list[Any]): The quantum tape operation parameters. If not provided,
+                the current tape parameter values are used (via :meth:`~.get_parameters`).
+        """
+        device.reset()
```
Comment (co9olguy) [inline pennylane/beta/tapes/tape.py:902]: will this introduce overhead for every execution? is it unavoidable?

File: pennylane/beta/tapes/tape.py [hunk 6]
```diff
@@ -833,3 +835,105 @@ def data(self):
...
+
+        # backup the current parameters
+        current_parameters = self.get_parameters()
```
Comment (co9olguy) [inline pennylane/beta/tapes/tape.py:905]: matbe use another term than "current"? ("cached"? "saved"? etc)

File: pennylane/beta/tapes/tape.py [hunk 7]
```diff
@@ -833,3 +835,105 @@ def data(self):
...
+
+        # temporarily mutate the in-place parameters
+        self.set_parameters(params)
+
+        if isinstance(device, qml.QubitDevice):
+            res = device.execute(self)
+        else:
+            res = device.execute(self.operations, self.observables, {})
```
Comment (co9olguy) [inline pennylane/beta/tapes/tape.py:913]: We're almost there :laughing:

File: tests/beta/tapes/test_tape.py [hunk 8]
```diff
@@ -709,3 +709,288 @@ def test_measurement_expansion(self):
...
+            qml.RX(params[0], wires=[0])
+            qml.RY(params[1], wires=[1])
+
+        res = tape.execute(dev)
+        assert res.size == 0
+        assert np.all(res == np.array([]))
+
+    def test_incorrect_output_dim_estimate(self):
+        """Test that a quantum tape with an incorrect output dimension
+        estimate corrects itself after evaluation."""
```
Comment (co9olguy) [inline tests/beta/tapes/test_tape.py:761]: I have to admit the term "estimate" still confuses me in this context :thinking:

File: tests/beta/tapes/test_tape.py [hunk 9]
```diff
@@ -709,3 +709,288 @@ def test_measurement_expansion(self):
...
+        dev = qml.device("default.qubit", wires=3)
+        params = [1.0, 1.0, 1.0]
+
+        with QuantumTape() as tape:
+            qml.RX(params[0], wires=[0])
+            qml.RY(params[1], wires=[1])
+            qml.RZ(params[2], wires=[2])
+            qml.CNOT(wires=[0, 1])
+            probs(wires=0)
+            probs(wires=[1])
```
Comment (co9olguy) [inline tests/beta/tapes/test_tape.py:771]: Any reason you used `0` and `[1]`?

File: tests/beta/tapes/test_tape.py [hunk 10]
```diff
@@ -709,3 +709,288 @@ def test_measurement_expansion(self):
...
+
+        assert isinstance(res[0], float)
+        assert np.allclose(res[0], np.cos(x), atol=tol, rtol=0)
+
+        assert isinstance(res[1], np.ndarray)
+        assert np.allclose(res[1], np.abs(dev.state) ** 2, atol=tol, rtol=0)
+
+    def test_single_mode_sample(self):
+        """Test that there is only one array of values returned
+        for single mode samples"""
```
Comment (co9olguy) [inline tests/beta/tapes/test_tape.py:896]: What is a single "mode" here?

File: tests/beta/tapes/test_tape.py [hunk 11]
```diff
@@ -709,3 +709,288 @@ def test_measurement_expansion(self):
...
+            qml.CNOT(wires=[0, 1])
+            sample(qml.PauliZ(0))
+            expval(qml.PauliZ(1))
+
+        res = tape.execute(dev)
+        assert res[0].shape == (10,)
+        assert isinstance(res[1], float)
+
+    def test_decomposition(self, tol):
+        """Test decomposition onto a devices supported gate set"""
```
Comment (co9olguy) [inline tests/beta/tapes/test_tape.py:946]:
```suggestion
        """Test decomposition onto a device's supported gate set"""
```

File: tests/beta/tapes/test_tape.py [hunk 12]
```diff
@@ -709,3 +709,288 @@ def test_measurement_expansion(self):
...
+            expval(qml.NumberOperator(0))
+
+        assert tape.output_dim == 1
+
+        res = tape.execute(dev)
+        assert res.shape == (1,)
+
+    def test_multiple_output_values(self, tol):
+        """Tests correct output shape and evaluation for a tape
+        with multiple expval outputs"""
```
Comment (co9olguy) [inline tests/beta/tapes/test_tape.py:981]: test covers a slightly different case than the docstring

File: tests/beta/tapes/test_tape.py [hunk 13]
```diff
@@ -709,3 +709,288 @@ def test_measurement_expansion(self):
...
+            qml.Displacement(x, 0, wires=[0])
+            qml.Squeezing(y, 0, wires=[1])
+            qml.Beamsplitter(np.pi / 4, 0, wires=[0, 1])
+            expval(qml.PolyXP(np.diag([0, 1, 0]), wires=0))  # X^2
+            var(qml.P(1))
+
+        assert tape.output_dim == 2
+
+        res = tape.execute(dev)
+        assert res.shape == (2,)
```
Comment (co9olguy) [inline tests/beta/tapes/test_tape.py:996]: Nice tests!

**Review by co9olguy — APPROVED** [review]
Thanks @josh146, looks great

File: pennylane/beta/tapes/tape.py [hunk 14]
```diff
@@ -833,3 +835,105 @@ def data(self):
...
+
+        # Update output dim if incorrect.
```
Comment (antalszava) [inline pennylane/beta/tapes/tape.py:915]: How come that in certain cases the output dim needs to be updated? Might be worth noting here perhaps.

File: tests/beta/tapes/test_tape.py [hunk 15]
```diff
@@ -709,3 +709,288 @@ def test_measurement_expansion(self):
...
+        assert tape.output_dim == sum([2, 2])
+
+        # modify the output dim
+        tape._output_dim = 2
+
+        res = tape.execute(dev)
+        assert tape.output_dim == sum([2, 2])
+
+    def test_incorrect_ragged_output_dim_estimate(self):
+        """Test that a quantum tape with an incorrect *ragged* output dimension
```
Comment (antalszava) [inline tests/beta/tapes/test_tape.py:783]: Not completely sure about the distinction between this test case and the previous one. Perhaps might be worth parametrizing?

**Review by antalszava — APPROVED** [review]
Looks good to me! :100: Nice @josh146 :blush:

File: pennylane/beta/tapes/tape.py [hunk 16]
```diff
@@ -17,6 +17,8 @@
 # pylint: disable=too-many-instance-attributes,protected-access,too-many-branches
 import contextlib
 
+import numpy as np
```
[the remark this answers is not in the document]
Comment (josh146) [reply pennylane/beta/tapes/tape.py:20]: Oh! I think this can be solved by instead doing `qml.beta.queuing.MeasurementProcess`.

File: pennylane/beta/tapes/tape.py (hunk 5 above)
Comment (josh146) [reply pennylane/beta/tapes/tape.py:902]: That's a good question.

With the way devices currently work, I _think_ this in unavoidable (this is how the current QNode execution + device API works, so I didn't want to make a logical change here). But this is worth exploring later on, especially in the context of caching the pre-rotated state and re-using it.

File: pennylane/beta/tapes/tape.py (hunk 3 above)
Comment (josh146) [reply pennylane/beta/tapes/tape.py:930]: Woah! This in-line commenting by codecov is new

File: tests/beta/tapes/test_tape.py (hunk 8 above)
Comment (josh146) [reply tests/beta/tapes/test_tape.py:761]: 'infers' is a better term, I'll change it.

File: tests/beta/tapes/test_tape.py (hunk 9 above)
Comment (josh146) [reply tests/beta/tapes/test_tape.py:771]: Nope, just conscious about using different patterns throughout the tests (I don't want to leave out edge cases)

File: tests/beta/tapes/test_tape.py (hunk 15 above)
Comment (josh146) [reply tests/beta/tapes/test_tape.py:783]: The difference is in the tape measurements 🙂  The previous test had output `[probs(wires=0), probs(wires=1)]` which will output a NumPy array with shape `(2,2)`.

This test instead has a tape with output `[probs(wires=0), probs(wires=[1, 2])]`, which will be a _ragged_ NumPy object array that looks like `[[a, b], [c, d, e, f]]`. So we are dealing with a completely different `dtype` to the previous case.

Since we are testing different logic here, I don't think this should be parametrized (parametrization should be restricted to testing the same logic with different data).

File: tests/beta/tapes/test_tape.py (hunk 10 above)
Comment (josh146) [reply tests/beta/tapes/test_tape.py:896]: Oops 🤦 
```suggestion
        for single wires samples"""
```

File: tests/beta/tapes/test_tape.py (hunk 12 above)
Comment (josh146) [reply tests/beta/tapes/test_tape.py:981]:
```suggestion
        with multiple measurement types
```

File: pennylane/beta/tapes/tape.py [hunk 17]
```diff
@@ -833,3 +835,105 @@ def data(self):
     @data.setter
     def data(self, params):
         self.set_parameters(params, trainable_only=False)
+
+    # ========================================================
+    # execution methods
+    # ========================================================
+
+    def execute(self, device, params=None):
```
[the remark this answers is not in the document]
Comment (josh146) [reply pennylane/beta/tapes/tape.py:843]: I have to admit, I'm a bit personally biased towards `execute` over `evaluate`. Partially because I find `execute` is more descriptive in that it will be actually calling the quantum device for the result.

Having a different term here will also make the final find-all/replace-all easier when we replace core!

But happy to change it back if everyone agrees

File: tests/beta/tapes/test_tape.py (hunk 8 above)
Comment (josh146) [inline tests/beta/tapes/test_tape.py:761]:
```suggestion
        """Test that a quantum tape with an incorrect inferred output dimension
        corrects itself after evaluation."""
```

File: pennylane/beta/tapes/tape.py (hunk 6 above)
Comment (josh146) [reply pennylane/beta/tapes/tape.py:905]:
```suggestion
        saved_parameters = self.get_parameters()
```

File: pennylane/beta/tapes/tape.py [hunk 18]
```diff
@@ -833,3 +835,105 @@ def data(self):
...
+            if self.output_dim != output_dim:
+                # update the output dimension estimate with the correct value
+                self._output_dim = output_dim
+
+        except (AttributeError, TypeError):
+            # unable to determine the output dimension
+            pass
+
+        # restore original parameters
+        self.set_parameters(current_parameters)
```
Comment (josh146) [inline pennylane/beta/tapes/tape.py:933]:
```suggestion
        self.set_parameters(saved_parameters)
```

**Comment by mariaschuld** [comment]
> > @mariaschuld pointed out that there are some potential advantages to instead having
> 
> Is there a link to the conversation? I'm curious what the advantages are.

Hey @trbromley ...this came from the idea of batching: if you want to send a list of tapes, you could do 

``` python
dev.execute([tape1, tape2,...])
```

But my latest favourite solution for batching would be to design a tape so that it can represent a batch of tapes itself, in which case ``tape.execute()`` should work fine!

File: pennylane/beta/tapes/tape.py (hunk 14 above)
Comment (josh146) [reply pennylane/beta/tapes/tape.py:915]: The output dimension of the QNode needs to be known in order for the Jacobian to be created properly. Currently, the QNode computes the output dimension when processing the queue:

* If a measurement process that returns a scalar is included, the output dim is incremented by 1
* If probability is included, the output dim is incremented by `2**len(wires)`
* If a CV probability is included, the output dim is incremented by `cutoff ** len(wires)`.

However, this is quite inelegent. It means that devices aren't free to define their own output without needing to always make the QNode aware of what the output dimension is (example: when we added CV probability, we needed to update the output dim in the QNode).

Instead, what we can do is add another approach; the tape simply looks out the output size after execution, without needing to know anything in advance!

This has a drawback though --- you always need to evaluate the QNode before computing the Jacobian.
